modules/azure/client_factory.go

```diff
@@ -0,0 +1,147 @@
+rather, additional clients will me added as-needed.
+
+*/
+
+package azure
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2019-07-01/compute"
+	"github.com/Azure/azure-sdk-for-go/services/containerservice/mgmt/2019-11-01/containerservice"
+	"github.com/Azure/azure-sdk-for-go/services/resources/mgmt/2019-06-01/subscriptions"
+	autorestAzure "github.com/Azure/go-autorest/autorest/azure"
+)
+
+const (
+	// AzureEnvironmentEnvName is the name of the Azure environment to use. Set to one of the following:
+	//
+	// "AzureChinaCloud":        ChinaCloud
+	// "AzureGermanCloud":       GermanCloud
+	// "AzurePublicCloud":       PublicCloud
+	// "AzureUSGovernmentCloud": USGovernmentCloud
+	// "AzureStackCloud":		 Azure stack
+	AzureEnvironmentEnvName = "AZURE_ENVIRONMENT"
+)
+
+// ClientType describes the type of client a module can create.
+type ClientType int
+
+const (
+	// SubscriptionsClientType represents a SubscriptionClient
+	SubscriptionsClientType ClientType = iota
+
+	// VirtualMachinesClientType represents a VirtualMachinesClient
+	VirtualMachinesClientType
+
+	// ManagedClustersClientType represents a ManagedClustersClient
+	ManagedClustersClientType
+)
+
+// ClientFactory describes the methods available on client factory implementatoins
+type ClientFactory interface {
+	// GetClientE returns a client instance based on the ClientType passed, or optionally an error.
+	GetClientE(clientType ClientType, subscriptionID string) (interface{}, error)
+
+	// GetClientBaseURIE returns the configured BaseURI for the given client, based on the passed ClientType.
+	GetClientBaseURIE(clientType ClientType, client interface{}) (string, error)
+}
+
+// multiEnvClientFactory is used to coordinate handing out properly configured Azure SDK clients
+// that are properly setup for use with Public or Sovereign clouds (depending on configuration)
+type multiEnvClientFactory struct{}
+
+// NewClientFactory returns a new multi-environment client factory
+func NewClientFactory() ClientFactory {
+	return &multiEnvClientFactory{}
+}
+
+// GetClientE returns a client instance based on the ClientType passed, or optionally an error.
+func (factory *multiEnvClientFactory) GetClientE(clientType ClientType, subscriptionID string) (interface{}, error) {
+	// Validate Azure subscription ID
+	subscriptionID, err := getTargetAzureSubscription(subscriptionID)
+	if err != nil {
+		return nil, err
+	}
+
+	// Lookup environment URI
+	baseURI, err := factory.getEnvironmentBaseURIE()
+	if err != nil {
+		return nil, err
+	}
+
+	// Create correct client based on type passed
+	switch clientType {
+	case SubscriptionsClientType:
+		return subscriptions.NewClientWithBaseURI(baseURI), nil
+	case VirtualMachinesClientType:
+		return compute.NewVirtualMachinesClientWithBaseURI(baseURI, subscriptionID), nil
+	case ManagedClustersClientType:
+		return containerservice.NewManagedClustersClientWithBaseURI(baseURI, subscriptionID), nil
+	}
```

Comment:
I found the need to cast these clients quite painful, adding several more error-prone lines to the API... I know Go doesn't have generics yet, so there are limits to what we can do here, but perhaps it would be cleaner to have a `GetClientXXX` method for each client type, and that method can return the exact type users ask for? E.g., `GetClientVirtualMachines`, `GetClientManagedClusters`, etc. This would result in a tiny bit of extra copy/paste within `client_factory.go`, but it would significantly reduce the boilerplate for all users, and remove the need for casting.

Reply:
Added explicit methods for each client type and removed general case / casting need.